### ISSUE TITLE
[Request] Promisify Debounce

### ISSUE THREAD

**Issue by jsfeldman** [quasarframework/quasar#1389]
I think it would be useful to promisify the [debounce function](http://quasar-framework.org/components/other-utils.html#Debounce-Function) so that the return value of the function being debounced can be accessed. The change in the code should be pretty minor and I think it would add a lot of functionality. Also, it shouldn't break anything unless people are relying on the debounce function to return undefined. Below is the lightly modified debounce function that I think should work:

```
export function debounce (fn, wait = 250, immediate) {
  let timeout
  let rej

  function debounced (...args) {
    return new Promise((resolve, reject) => {
      rej = reject
      const later = () => {
        timeout = null
        if (!immediate) {
          resolve(fn.apply(this, args))
        }
      }

      clearTimeout(timeout)
      if (immediate && !timeout) {
        fn.apply(this, args)
      }
      timeout = setTimeout(later, wait)
    })
  }

  debounced.cancel = () => {
    clearTimeout(timeout)
    rej('debounce canceled')
  }

  return debounced
}
```

The use case I currently need it for is debouncing [asyncronous validations](https://monterail.github.io/vuelidate/#sub-asynchronous-validation) in vuelidate since, at least on the version of vuelidate I'm using, using the delayed validation errors pattern doesn't stop the asyncronous rule from running. But I'm sure there are a lot of other use cases out there as well.

**Comment by pdanpdan** [comment]
As per vuelidate's own suggestion, debouncing should be done before the validation function is called.

For the debounced promise I think there are some implementations available in the wild, and they are pretty different in the way they behave, so I think this should be the user's choice whick one to use.

**Comment by jsfeldman** [comment]
Hmm, the only documentation I could find on vuelidate about debouncing is [delayed validation errors](https://monterail.github.io/vuelidate/#sub-delayed-validation-errors), which, based on the example, looks like the validation is run whenever you make a change and just doesn't set $dirty, or $error until the specified time has passed.

The only way I can think of debouncing before the validation function is called is if you debounced when setting the value in the v-model. I guess you could achieve that by using :value and @input instead of v-model, and make the function @input calls a debounced one. But since vuelidate already has functionality for working with promises in the validation, and you might not want the field itself to be debounced, just the validation, I thought this would be simpler. Is there another way to debounce before validation I'm not thinking of?

In terms of the idea of debounced promises in general, I was just thinking that a debounced promises is essentially a superset of the current implementation and would be nice to have. It would save users time finding and implementing another implementation with little change to the codeebase.

**Comment by pdanpdan** [comment]
Yes, debounce setting the model.
From vuelidate:

> Validator is evaluated on every data change, as it is essentially a computed value. If you need to throttle an async call, do it on your data change event, not on the validator itself. You may end up with broken Vue observables otherwise.

**Comment by jsfeldman** [comment]
Ahh, thanks! I completely glossed over that section. I always feel a little uneasy when the v-model value is not synced with the input field value, but i guess they have their reasons.

**Comment by stefanvanherwijnen** [comment]
I have a component which requires data from a Vuex store. If this data is not available, the data is retrieved through an axios request. However, if multiple components are initialized without the data being set, multiple requests are made. I solved this by using debounce, but it is not possible to wait on the request (i.e. use then()).
The function provided by @jsfeldman seems to fix this by returning the Promise inside the debounce function. Is there a reason not to do it like this? (Or, what is the reason for not implementing this debounce function?)
@rstoenescu

**Comment by rstoenescu** [comment]
@stefanvanherwijnen Hi,

Because the `debounce` method should only return a debounced function, not a Promise. The most use cases (but not all) for `debounce` are debouncing component methods. It's not trying to fit any custom scenarios, just doing one job and one job only. However, if you need the implementation above, you can just copy paste it into a file and import it where needed.

**Comment by stefanvanherwijnen** [comment]
Thanks Razvan, that makes sense. I already copied the function and that works fine for my specific case. Now I think of it, it is probably better indeed not to implement workarounds for specific use cases in the source code, even though it would make no difference for functionality.

So for people that need this functionality: just copy the function provided by jsfeldman :)
